kxclib.ml

```diff
@@ -1622,7 +1667,82 @@ end = struct
-  let to_jsonm = failwith "noimpl: to_jsonm"
+  let of_jsonm' : 'loc jsonm' -> (jv, 'loc) result = fun jsonm' ->
+    let (>>=) = Result.bind in
+    let of_jsonm'_token : 'loc * jsonm_token -> (jv, 'loc) result = function
+      | _, `Null -> Ok `null
+      | _, `Bool x -> Ok (`bool x)
+      | _, `String x -> Ok (`str x)
+      | _, `Float x -> Ok (`num x)
+      | loc, _ -> Error loc in
+    let rec collect_arr : 'loc jsonm' -> ((jv * 'loc jsonm'), 'loc) result = fun jsonm' ->
+      let rec aux : 'loc jsonm' -> (jv list, 'loc) result -> ((jv * 'loc jsonm'), 'loc) result = fun seq acc ->
+        acc >>= fun acc ->
+        match seq () with
+        | Seq.Nil -> raise End_of_file
+        | Seq.Cons ((_, `Ae), rest) ->
+          Ok (`arr (List.rev acc), rest)
+        | Seq.Cons ((_, `As), rest) ->
+          collect_arr rest >>= fun (arr, rest) ->
+          aux rest (Ok (arr :: acc))
+        | Seq.Cons ((_, `Os), rest) ->
+          collect_obj rest >>= fun (obj, rest) ->
+          aux rest (Ok (obj :: acc))
+        | Seq.Cons (tok', rest) ->
+          of_jsonm'_token tok' >>= fun x ->
+          aux rest (Ok (x :: acc)) in
+      aux jsonm' (Ok [])
+    and collect_obj : 'loc jsonm' -> ((jv * 'loc jsonm'), 'loc) result = fun jsonm' ->
+      let rec aux : 'loc jsonm' -> ((string * jv) list, 'loc) result -> ((jv * 'loc jsonm'), 'loc) result = fun seq acc ->
+          acc >>= fun acc ->
+          match seq () with
+          | Seq.Nil -> raise End_of_file
+          | Seq.Cons ((_, `Oe), rest) ->
+            Ok (`obj (List.rev acc), rest)
+          | Seq.Cons ((_, `Name name), rest) -> begin match rest () with
+              | Seq.Nil -> raise End_of_file
+              | Seq.Cons ((_, `As), rest) ->
+                collect_arr rest >>= fun (arr, rest) ->
+                aux rest (Ok ((name, arr) :: acc))
+              | Seq.Cons ((_, `Os), rest) ->
+                collect_obj rest >>= fun (obj, rest) ->
+                aux rest (Ok ((name, obj) :: acc))
+              | Seq.Cons (tok', rest) ->
+                of_jsonm'_token tok' >>= fun x ->
+                aux rest (Ok ((name, x) :: acc))
+            end
+          | Seq.Cons ((loc, _), _) -> Error loc in
+      aux jsonm' (Ok []) in
+    match jsonm' () with
+    | Seq.Cons (tok', rest) when rest () = Seq.Nil ->
+      of_jsonm'_token tok'
+    | Seq.Cons ((loc, `As), rest) ->
+      collect_arr rest >>= fun (arr, rest) -> begin match rest () with
+        | Seq.Nil -> Ok arr
+        | _ -> Error loc
+      end
+    | Seq.Cons ((loc, `Os), rest) ->
+      collect_obj rest >>= fun (obj, rest) -> begin match rest () with
+        | Seq.Nil -> Ok obj
+        | _ -> Error loc
+      end
+    | Seq.Cons ((loc, _), _) -> Error loc
+    | Seq.Nil -> raise End_of_file
+  let of_jsonm : jsonm -> jv option = fun jsonm ->
+    Seq.map (fun tok -> ((), tok)) jsonm
+    |> of_jsonm'
+    |> Result.to_option
+  let rec to_jsonm : jv -> jsonm = function
```

Comment:
pls add `(* XXX - optimize *)` since this is a quite suboptimal implementation
(good enough for now, but might not fits the needs of many future projects, especially those deliberately chose to use jsonm for performance reason)